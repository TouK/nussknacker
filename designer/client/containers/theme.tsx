/* eslint-disable i18next/no-literal-string */
import {css} from "@emotion/css"
import {Global, Theme, ThemeProvider, ThemeProviderProps, useTheme} from "@emotion/react"
import Color from "color"
import React, {useMemo} from "react"
import {darkTheme} from "./darkTheme"
import {defaultAppTheme} from "./defaultAppTheme"

type DeepPartial<T> = {
  [P in keyof T]?: DeepPartial<T[P]>;
}

export function tint(base: string, amount = 0): string {
  return Color(base).mix(Color("white"), amount).hsl().string()
}

export function alpha(base: string, amount = 1): string {
  return Color(base).alpha(amount).hsl().string()
}

export function tintPrimary(base: string): {
  primary75: string,
  primary25: string,
  primary50: string,
  primary: string,
} {
  return {
    primary: tint(base, 0),
    primary75: tint(base, 0.75),
    primary50: tint(base, 0.5),
    primary25: tint(base, 0.25),
  }
}

export type NkTheme = DeepPartial<typeof defaultAppTheme>

declare module "@emotion/react" {
  // eslint-disable-next-line @typescript-eslint/no-empty-interface
  interface Theme extends NkTheme {
  }
}

export function NkThemeProvider({theme = darkTheme, ...props}: Partial<ThemeProviderProps>): JSX.Element {
  return <ThemeProvider theme={theme} {...props}/>
}

export const useNkTheme: () => { withFocus: string, theme: NkTheme } = () => {
  const theme = useTheme()

  const withFocus = useMemo(() => css({
    ":focus, :active:focus": {
      outline: "none",
      borderColor: theme?.colors?.focusColor,
      boxShadow: `0 0 0 1px ${theme?.colors?.focusColor}`,
    },
  }), [theme])

  return {theme, withFocus}
}

export function getContrastColor(color: string, contrast = 5, ratio = .1) {
  const baseColor = Color(color)
  const resultColor = baseColor.isDark() ? baseColor.lighten(ratio) : baseColor.darken(ratio)
  if (baseColor.contrast(resultColor) < contrast && resultColor.lightness() % 100 !== 0) {
    return getContrastColor(color, contrast, ratio * 1.1)
  }
  return resultColor.hex()
}

export function getDarkenContrastColor(color: string, contrast = 5, ratio = .1) {
  const baseColor = Color(color)
  const resultColor = baseColor.darken(ratio)
  if (baseColor.contrast(resultColor) < contrast && resultColor.lightness() > 0) {
    return getDarkenContrastColor(color, contrast, ratio * 1.1)
  }
  return resultColor.hex()
}

function parseVarName(name: string) {
  return name.startsWith("--") ? name : `--${name}`
}

function parseColorName(name: string) {
  return name.toLowerCase().endsWith("color") ? name : `${name}Color`
}

function colorsToVariables(colors: Record<string, string>): Record<`--${string}`, string> {
  return Object.fromEntries(Object.entries(colors).map(
    ([name, value]) => [parseVarName(parseColorName(name)), value]
  ))
}

function themeToVariables(theme: Theme): { ":root": Record<`--${string}`, string> } {
  return {
    ":root": colorsToVariables(theme.colors),
  }
}

export const GlobalCSSVariables = () => (
  <Global styles={themeToVariables}/>
)
