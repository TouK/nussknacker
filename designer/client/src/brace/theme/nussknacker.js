//based on https://github.com/thlorenz/brace/blob/master/theme/monokai.js

import "@fontsource/roboto-mono"
const nussknackerCssTheme = require("!raw-loader!./nussknacker.css").default

ace.define("ace/theme/nussknacker", ["require", "exports", "module", "ace/lib/dom"], function (acequire, exports, module) {
  exports.isDark = true
  exports.cssClass = "ace-nussknacker"
  exports.cssText = nussknackerCssTheme

  const dom = acequire("../lib/dom")
  dom.importCssString(exports.cssText, exports.cssClass)
})
