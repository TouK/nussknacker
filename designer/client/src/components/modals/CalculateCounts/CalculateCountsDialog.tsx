/* eslint-disable i18next/no-literal-string */
import { css, cx } from "@emotion/css";
import { WindowButtonProps, WindowContentProps } from "@touk/window-manager";
import React, { PropsWithChildren, useCallback, useMemo, useState } from "react";
import { useTranslation } from "react-i18next";
import { useDispatch, useSelector } from "react-redux";
import { fetchAndDisplayProcessCounts } from "../../../actions/nk";
import { getProcessName, getScenarioGraph } from "../../../reducers/selectors/graph";
import { WindowContent } from "../../../windowManager";
import { PickerInput } from "./Picker";
import { CalculateCountsForm } from "./CalculateCountsForm";
import moment from "moment";
import { LoadingButtonTypes } from "../../../windowManager/LoadingButton";

export type State = {
    from: PickerInput;
    to: PickerInput;
};

const initState = (): State => {
    return {
        from: moment().startOf("day"),
        to: moment().endOf("day"),
    };
};

export function CountsDialog({ children, ...props }: PropsWithChildren<WindowContentProps>): JSX.Element {
    const { t } = useTranslation();
    const [state, setState] = useState(initState);
    const processName = useSelector(getProcessName);
    const dispatch = useDispatch();
    const scenarioGraph = useSelector(getScenarioGraph);

    const confirm = useCallback(async () => {
        await dispatch(fetchAndDisplayProcessCounts(processName, moment(state.from), moment(state.to), scenarioGraph));
    }, [dispatch, processName, state.from, state.to, scenarioGraph]);

    const isStateValid = moment(state.from).isValid() && moment(state.to).isValid();
    const buttons: WindowButtonProps[] = useMemo(
        () => [
            {
                title: t("dialog.button.cancel", "Cancel"),
                action: () => {
                    props.close();
                },
                classname: LoadingButtonTypes.secondaryButton,
            },
            {
                title: t("dialog.button.ok", "Ok"),
                disabled: !isStateValid,
                action: async () => {
                    await confirm();
                    props.close();
                },
            },
        ],
        [confirm, isStateValid, props, t],
    );

    return (
        <WindowContent
            buttons={buttons}
            title={t("calculateCounts.title", "counts")}
            classnames={{
                content: cx(
                    "modalContentDark",
                    css({
                        padding: "0 2em 2em",
                        textAlign: "center",
                        p: {
                            marginTop: "30px",
                        },
                    }),
                ),
            }}
            {...props}
        >
            <CalculateCountsForm value={state} onChange={setState} />
        </WindowContent>
    );
}
