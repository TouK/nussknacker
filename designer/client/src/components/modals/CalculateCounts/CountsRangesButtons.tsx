import moment, { Moment } from "moment";
import React, { PropsWithChildren, useCallback, useMemo } from "react";
import { PredefinedDropdownButton, PredefinedRangeButton } from "./CountsStyled";
import { useTranslation } from "react-i18next";

export interface Range {
    name: string;
    from: () => Moment;
    to: () => Moment;
    refresh?: number;
}

interface RangesButtonsProps {
    ranges: Range[];
    onChange: (value: [Moment, Moment], refresh?: number | null) => void;
    limit?: number;
}

export function CountsRangesButtons({ children, ranges, onChange, limit = -1 }: PropsWithChildren<RangesButtonsProps>): JSX.Element {
    const { t } = useTranslation();
    const changeHandler = useCallback(
        (range: Range) => {
            const from = range.from();
            const to = range.to();
            const refresh = range.refresh ?? (moment().isBefore(to) ? 10 : null);
            onChange([from, to], refresh);
        },
        [onChange],
    );

    const visible = useMemo(() => (limit >= 0 ? ranges.slice(0, limit) : ranges), [ranges, limit]);
    const collapsed = useMemo(() => (limit >= 0 ? ranges.slice(limit) : []), [ranges, limit]);
    return (
        <>
            {visible.map((range) => (
                <PredefinedRangeButton
                    key={range.name}
                    type="button"
                    title={range.name}
                    onClick={() => changeHandler(range)}
                    style={{ flex: 1 }}
                >
                    {range.name}
                </PredefinedRangeButton>
            ))}

            {collapsed.length > 0 ? (
                <PredefinedDropdownButton
                    options={collapsed.map((value) => ({ label: value.name, value }))}
                    onRangeSelect={changeHandler}
                    style={{ flex: 1 }}
                    wrapperStyle={{
                        display: "flex",
                        flex: 2,
                    }}
                >
                    {children || t("calculateCounts.more", "Select more...")}
                </PredefinedDropdownButton>
            ) : null}
        </>
    );
}
