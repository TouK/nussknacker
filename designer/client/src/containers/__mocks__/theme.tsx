/* eslint-disable i18next/no-literal-string */
import React, { PropsWithChildren } from "react";

export const tintPrimary = jest.fn();

export const NkThemeProvider = ({ children }: PropsWithChildren<unknown>) => <>{children}</>;

export const useNkTheme = () => ({ theme: {}, withFocus: "withFocus" });
