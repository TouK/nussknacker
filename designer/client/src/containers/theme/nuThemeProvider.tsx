import { ThemeProvider } from "@mui/material/styles";
import React, { PropsWithChildren } from "react";
import { GlobalCSSVariables } from "./helpers";
import { nuTheme } from "./nuTheme";

export const NuThemeProvider: React.FC<PropsWithChildren<unknown>> = ({ children }) => {
    return (
        <ThemeProvider theme={nuTheme}>
            <GlobalCSSVariables />
            {children}
        </ThemeProvider>
    );
};
