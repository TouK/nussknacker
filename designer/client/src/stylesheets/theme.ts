import { Color } from "@mui/material";
import { createTheme } from "@mui/material/styles";

declare module "@mui/material/styles" {
    interface Palette {
        dark?: Partial<Color>;
        blue?: Partial<Color>;
        green?: Partial<Color>;
    }
    interface PaletteOptions {
        dark?: Partial<Color>;
        blue?: Partial<Color>;
        green?: Partial<Color>;
    }
}

let newTheme = createTheme({
    palette: {
        grey: {
            50: "#F5F5F5",
            100: "#E0E0E0",
            200: "#b3b3b3",
            300: "#777",
            400: "#666",
            500: "#4d4d4d",
            600: "#434343",
            700: "#333",
            800: "#3e3e3e",
            900: "#222",
            A100: "#f6f6f6",
            A200: "#999999",
            A400: "#808080",
            A700: "#616161",
        },
        blue: {
            200: "#0E9AE0",
            500: "#0058a9",
        },
        green: {
            200: "8fad60",
        },
        dark: {
            200: "#afafaf",
            300: "#444",
            500: "#ccc",
            600: "#5d5d5d",
            700: "#3a3a3a",
        },
    },
});

newTheme = createTheme(newTheme, {
    palette: {
        mode: "dark",
        primary: {
            main: newTheme.palette.blue[500], // Primary color
        },
        info: {
            main: newTheme.palette.dark[500], // Info color
        },
        warning: {
            main: "#FF9A4D", // Warning color
        },
        error: {
            main: "#f25c6e", // Error color
        },
        success: {
            main: "#64d864", // Success color
        },
    },
});
