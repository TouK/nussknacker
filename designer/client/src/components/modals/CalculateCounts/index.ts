import {CountsDialog} from "./CalculateCountsDialog"

export * from "./CalculateCountsDialog"

export default CountsDialog
