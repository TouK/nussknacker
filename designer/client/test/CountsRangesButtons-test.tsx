import moment from "moment";
import React from "react";
import { CountsRangesButtons } from "../src/components/modals/CalculateCounts/CountsRangesButtons";
import { fireEvent, render, screen } from "@testing-library/react";
import { jest } from "@jest/globals";
import { NuThemeProvider } from "../src/containers/theme/nuThemeProvider";

jest.mock("react-i18next", () => ({
    useTranslation: () => ({
        t: (key) => key,
        i18n: { changeLanguage: () => {} },
    }),
}));

describe("CountsRangesButtons tests", () => {
    const m = moment("2001-10-19T23:00:00.000Z");
    const range1 = { name: "range1", from: () => m.clone(), to: () => m.clone().add(1, "hour") };
    const range2 = { name: "range2", from: () => m.clone().add(1, "day"), to: () => m.clone().add(2, "days") };
    const range3 = { name: "range3", from: () => m.clone().add(1, "week"), to: () => m.clone().add(2, "weeks") };
    const ranges = [range1, range2, range3];
    const changeFn = jest.fn();

    beforeEach(() => {
        changeFn.mockReset();
    });

    it("should render buttons", () => {
        const { container } = render(
            <NuThemeProvider>
                <CountsRangesButtons ranges={ranges} onChange={changeFn} limit={2} />
            </NuThemeProvider>,
        );
        expect(container).toMatchSnapshot();
    });

    it("should handle click", () => {
        render(
            <NuThemeProvider>
                <CountsRangesButtons ranges={ranges} onChange={changeFn} />
            </NuThemeProvider>,
        );
        fireEvent.click(screen.getByRole("button", { name: /range1/ }));
        expect(changeFn).toHaveBeenCalledTimes(1);
        expect(changeFn).toHaveBeenCalledWith([range1.from(), range1.to()], null);
    });

    it("should collapse buttons", () => {
        render(
            <div>
                <NuThemeProvider>
                    <CountsRangesButtons ranges={ranges} onChange={changeFn} limit={1} />
                </NuThemeProvider>
            </div>,
        );

        const buttons = screen.getAllByRole("button");
        expect(buttons).toHaveLength(2);
        const options = document.getElementsByClassName("nodeValueSelect__option");
        expect(options).toHaveLength(0);
        fireEvent.click(buttons[buttons.length - 1]);
        const options2 = document.getElementsByClassName("nodeValueSelect__option");
        expect(options2).toHaveLength(2);
        fireEvent.click(options2.item(1));
        expect(changeFn).toHaveBeenCalledTimes(1);
        expect(changeFn).toHaveBeenCalledWith([range3.from(), range3.to()], null);
    });

    it("should hide expand button when not needed", () => {
        render(
            <NuThemeProvider>
                <CountsRangesButtons ranges={ranges} onChange={changeFn} limit={10} />
            </NuThemeProvider>,
        );

        const buttons = screen.getAllByRole("button");
        expect(buttons).toHaveLength(3);
        fireEvent.click(buttons[buttons.length - 1]);
        expect(changeFn).toHaveBeenCalledTimes(1);
        expect(changeFn).toHaveBeenCalledWith([range3.from(), range3.to()], null);
    });
});
