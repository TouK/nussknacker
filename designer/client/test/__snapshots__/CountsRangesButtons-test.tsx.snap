// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountsRangesButtons tests should render buttons 1`] = `
.emotion-0 {
  font-family: Inter,-apple-system,BlinkMacSystemFont,"Segoe UI","Helvetica Neue",Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  font-weight: 400;
  font-size: 12px;
  line-height: 1.75;
  text-transform: capitalize;
  border-radius: 0;
  background-color: #242F3E;
  -webkit-transition: #242F3E 0.2s;
  transition: #242F3E 0.2s;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  min-width: 80px;
  padding: 5px;
  margin: 10px;
}

.emotion-0:disabled,
.emotion-0.disabled {
  opacity: 0.3;
  cursor: not-allowed!important;
}

.emotion-2 {
  font-family: Inter,-apple-system,BlinkMacSystemFont,"Segoe UI","Helvetica Neue",Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  font-weight: 400;
  font-size: 12px;
  line-height: 1.75;
  text-transform: capitalize;
  border-radius: 0;
  background-color: #242F3E;
  -webkit-transition: #242F3E 0.2s;
  transition: #242F3E 0.2s;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 100%;
  padding: 5px;
  margin: 10px;
}

.emotion-2:disabled,
.emotion-2.disabled {
  opacity: 0.3;
  cursor: not-allowed!important;
}

<div>
  <button
    class="emotion-0"
    style="flex: 1;"
    title="range1"
    type="button"
  >
    range1
  </button>
  <button
    class="emotion-0"
    style="flex: 1;"
    title="range2"
    type="button"
  >
    range2
  </button>
  <span
    style="display: flex; flex: 2; position: relative;"
  >
    <span
      style="position: relative; width: 100%; height: 100%; display: flex;"
    >
      <button
        class="emotion-2"
        style="flex: 1;"
        type="button"
      >
        calculateCounts.more
      </button>
    </span>
  </span>
</div>
`;
